File: Win32Lib/Protect.cpp
#include "stdafx.h"
#include "Protect.h"

#include <Windows.h>

#include "Win32Exception.h"

using namespace std;

namespace NeoEdit
{
	namespace Win32LibNS
	{
		namespace Processes
		{
			Protect::Protect(shared_ptr<void> handle, shared_ptr<const VirtualQueryInfo> info, uint32_t protect) :
				handle(handle),
				info(info),
				protect(protect)
			{
				if (protect == info->Protect)
					return;

				DWORD oldProtect;
				if (!VirtualProtectEx((HANDLE)handle.get(), info->StartAddress, info->RegionSize, protect, &oldProtect))
					Win32Exception::Throw();
			}

			Protect::~Protect()
			{
				if (protect == info->Protect)
					return;

				DWORD oldProtect;
				if (!VirtualProtectEx((HANDLE)handle.get(), info->StartAddress, info->RegionSize, info->Protect, &oldProtect))
					return;
				if (!FlushInstructionCache((HANDLE)handle.get(), info->StartAddress, info->RegionSize))
					return;
			}
		}
	}
}

File: HexEdit/Protect.cs
using System;
using System.ComponentModel;
using Microsoft.Win32.SafeHandles;

namespace NeoEdit.HexEdit
{
	class Protect : IDisposable
	{
		readonly SafeProcessHandle handle;
		readonly VirtualQueryInfo info;
		readonly int protect;

		public Protect(SafeProcessHandle handle, VirtualQueryInfo info, int protect)
		{
			this.handle = handle;
			this.info = info;
			this.protect = protect;

			if (protect == info.Protect)
				return;

			int oldProtect;
			if (!Win32.VirtualProtectEx(handle.DangerousGetHandle(), (IntPtr)info.StartAddress, (IntPtr)info.RegionSize, protect, out oldProtect))
				throw new Win32Exception();
		}

		public void Dispose()
		{
			if (protect == info.Protect)
				return;

			int oldProtect;
			if (!Win32.VirtualProtectEx(handle.DangerousGetHandle(), (IntPtr)info.StartAddress, (IntPtr)info.RegionSize, info.Protect, out oldProtect))
				return;
			if (!Win32.FlushInstructionCache(handle.DangerousGetHandle(), (IntPtr)info.StartAddress, (IntPtr)info.RegionSize))
				return;
		}
	}
}
